### DISCUSSION TITLE
How to integrate 'equivalent_slip_increment'

### DISCUSSION THREAD

**Discussion by pshen20127** [idaholab/moose#23365] [Q&A Modules: Solid mechanics] [answered]
Hello All

I am trying to integrate 'equivalent_slip_increment' in the file 'CrystalPlasticityStressUpdateBaseAddStateVars.C'. I added several lines as below:
```
106   equivalent_slip_increment(declareProperty<RankTwoTensor>("equivalent_slip_increment")),
107   equivalent_slip_increment_old(getMaterialPropertyOld<RankTwoTensor>("equivalent_slip_increment")),

501   equivalent_slip_increment += equivalent_slip_increment_old;
```
But it met an error like this:
![Screenshot from 2023-02-07 10-52-46](https://user-images.githubusercontent.com/81499506/217331486-4b0083d9-9fe2-40e1-b0ef-d4e0ca02ff0c.png)

What's wrong with it and how to fix it? Here is may codes:https://github.com/pshen20127/IN625.git

Thanks
Pengfei

**Comment by ajacquey** [comment, marked as answer]
Try:

`equivalent_slip_increment[_qp] += equivalent_slip_increment_old[_qp];`

**Reply by pshen20127** [reply]
Thank you for your reply. I have tried and it doesn't work. This is error message.
![Screenshot from 2023-02-07 14-15-26](https://user-images.githubusercontent.com/81499506/217367426-36fa8cb8-5557-4f54-b75e-7c38ebb6e4cc.png)

**Reply by hugary1995** [reply]
You probably are overriding equivalent_slip_increment locally... try
```
equivalent_slip_increment += equivalent_slip_increment_old[_qp];
```
which should work but indicate something else you are doing is wrong.

**Reply by ajacquey** [reply]
There are a number of things which are incorrect in your files:

- The material property `equivalent_slip_increment` is declared as a stateful property (you need the old value) but is not initialized in `initQpStatefulProperties()`
- The function where the the operation `equivalent_slip_increment[_qp] += equivalent_slip_increment_old[_qp];` takes place has `equivalent_slip_increment` as argument. I am not sure what you are doing because I did not find where and how this function is called in your file. In principle, `equivalent_slip_increment += equivalent_slip_increment_old[_qp];`  should work but it's confusing. I advise renaming your material property `_equivalent_slip_increment` to differentiate between local variable and material property.

**Reply by pshen20127** [reply]
I tried `equivalent_slip_increment += equivalent_slip_increment_old[_qp];` and it works. Could you please explain why the former `equivalent_slip_increment` doesn't need `[_qp]` but the latter one needs?

**Reply by hugary1995** [reply]
You are adding a rank two tensor to another. The LHS is a rank two tensor, the RHS also has to be a rank two tensor. Your `equivalent_slip_increment_old` is an array, hence the [] accessor to retrieve the rank two tensor inside.

Please follow @ajacquey 's suggestions above.

**Reply by pshen20127** [reply]
Thanks, will do. I also declare `equivalent_slip_increment_old` as a rank two tensor like below. 
```
equivalent_slip_increment_old(getMaterialPropertyOld<RankTwoTensor>("equivalent_slip_increment")),
```
Why do you say it is array?

**Reply by hugary1995** [reply]
`getMaterialPropertyOld<RankTwoTensor>("equivalent_slip_increment")` returns a `const MaterialProperty<RankTwoTensor> &`. A `MaterialProperty<T>` is essentially a `MooseArray<T>` at its core, and `MooseArray` is our implementation of resizable array.

**Comment by sapitts** [comment]
As an aside, I would recommend caution in adding an old equivalent slip increment to the current slip increment: the current slip increment value is used in `ComputeMultipleCrystalPlasticityStress` to calculate the plastic deformation gradient, see https://github.com/idaholab/moose/blob/198b8ca429115454bf7e56c78ec5004492136d1a/modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeMultipleCrystalPlasticityStress.C#L507-L514

One of the beauties of open source code is that you can make whatever change you like :-) In this case you might want to consider the downstream implications as you make your changes

**Reply by pshen20127** [reply]
Thanks, get it. I will create another variable assigned the value of equivalent slip increment.

**Reply by pshen20127** [reply]
Hello

As you suggested, I add some new variables and codes. The main calculation I want to do is from line 505 as below. 
````
505   equivalent_shearstrain = equivalent_slip_increment;
506   equivalent_shearstrain += equivalent_shearstrain_old;
507   equivalent_plastic_strain = sqrt(pow(equivalent_shearstrain(0,0),2)+pow(equivalent_shearstrain(0,1),2)+pow(equivalent_shearstrain(0,2),2)
508                                  +pow(equivalent_shearstrain(1,0),2)+pow(equivalent_shearstrain(1,1),2)+pow(equivalent_shearstrain(1,2),2)
509                                  +pow(equivalent_shearstrain(2,0),2)+pow(equivalent_shearstrain(2,1),2)+pow(equivalent_shearstrain(2,2),2));
````
The equation from line 507 to 509 is 
![WeChat Screenshot_20230208170955](https://user-images.githubusercontent.com/81499506/217680615-db3c3e39-bd62-44db-a174-858568112d2a.png)
When I compiled the codes, I met many errors:
![WeChat Screenshot_20230208171240](https://user-images.githubusercontent.com/81499506/217681040-b4135d7e-6bde-4edc-82ef-4e920134ba9c.png)

Could you please tell me how to fix it? Here is my codes : https://github.com/pshen20127/IN625.git
